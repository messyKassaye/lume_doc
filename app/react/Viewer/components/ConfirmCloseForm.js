import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { actions as formActions } from 'react-redux-form';

import { hideModal } from 'app/Modals/actions/modalActions';
import Modal from 'app/Layout/Modal';
import { Icon } from 'UI';
import { closePanel } from '../actions/uiActions';

export class ConfirmCloseForm extends Component {
  confirm() {
    this.props.hideModal('ConfirmCloseForm');
    this.props.closePanel();
    this.props.resetForm('documentViewer.sidepanel.metadata');
  }

  render() {
    if (!this.props.doc) {
      return false;
    }

    const { doc } = this.props;

    return (
      <Modal isOpen={!!doc} type="danger">
        <Modal.Body>
          <h4>Confirm</h4>
          <p>All changes will be lost, are you sure you want to proceed?</p>
        </Modal.Body>

        <Modal.Footer>
          <button
            type="button"
            className="btn btn-default cancel-button"
            onClick={() => this.props.hideModal('ConfirmCloseForm')}
          >
            <Icon icon="times" /> Cancel
          </button>
          <button
            type="button"
            className="btn btn-danger confirm-button"
            onClick={() => this.confirm()}
          >
            <Icon icon="trash-alt" /> Ok
          </button>
        </Modal.Footer>
      </Modal>
    );
  }
}

ConfirmCloseForm.propTypes = {
  hideModal: PropTypes.func,
  resetForm: PropTypes.func,
  closePanel: PropTypes.func,
  doc: PropTypes.object,
};

const mapStateToProps = state => ({ doc: state.modals.get('ConfirmCloseForm') });

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ closePanel, hideModal, resetForm: formActions.reset }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(ConfirmCloseForm);
