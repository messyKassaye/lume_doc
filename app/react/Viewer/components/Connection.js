import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { fromJS as Immutable } from 'immutable';
import { I18NLink } from 'app/I18N';
import { NeedAuthorization } from 'app/Auth';
import { Icon } from 'UI';

import ShowIf from 'app/App/ShowIf';
import { deleteReference } from 'app/Viewer/actions/referencesActions';
import {
  highlightReference,
  activateReference,
  selectReference,
} from 'app/Viewer/actions/uiActions';
import { Item } from 'app/Layout';
import { createSelector } from 'reselect';
import helpers from 'app/Documents/helpers';

const selectDoc = createSelector(
  s => s.documentViewer.targetDoc,
  s => s.documentViewer.doc,
  (targetDoc, doc) =>
    targetDoc.get('_id')
      ? helpers.performantDocToJSWithoutRelations(targetDoc)
      : helpers.performantDocToJSWithoutRelations(doc)
);

export class Connection extends Component {
  clickReference(reference) {
    if (this.props.readOnly) {
      return;
    }
    if (!this.props.targetDoc) {
      this.props.activateReference(
        reference,
        this.props.doc.defaultDoc.pdfInfo,
        this.props.referencesSection
      );
    }

    if (this.props.targetDoc && typeof reference.reference !== 'undefined') {
      this.props.selectReference(reference, this.props.doc.defaultDoc.pdfInfo);
    }
  }

  deleteReference(reference) {
    this.context.confirm({
      accept: () => {
        this.props.deleteReference(reference);
      },
      title: 'Confirm delete connection',
      message: 'Are you sure you want to delete this connection?',
    });
  }

  relationType(id) {
    const type = this.props.relationTypes.find(relation => relation.get('_id') === id);
    if (type) {
      return type.name;
    }
  }

  render() {
    const { reference } = this.props;
    let itemClass = '';
    const disabled = this.props.targetDoc && typeof reference.reference === 'undefined';

    if (this.props.highlighted) {
      itemClass = 'relationship-hover';
    }

    if (this.props.active) {
      itemClass = 'relationship-active';
    }

    if (this.props.active && this.props.targetDoc && this.props.targetRange) {
      itemClass = 'relationship-selected';
    }

    if (this.props.readOnly) {
      itemClass = '';
    }

    const doc = Immutable(reference.associatedRelationship.entityData);
    return (
      <Item
        onMouseEnter={this.props.highlightReference.bind(null, reference._id)}
        onMouseLeave={this.props.highlightReference.bind(null, null)}
        onClick={this.clickReference.bind(this, reference)}
        doc={doc}
        noMetadata
        className={`${itemClass} item-${reference._id} ${disabled ? 'disabled' : ''} ${
          this.props.readOnly ? 'readOnly' : ''
        }`}
        data-id={reference._id}
        additionalText={
          reference.associatedRelationship.reference
            ? reference.associatedRelationship.reference.text
            : null
        }
        additionalMetadata={[
          { label: 'Connection type', value: this.relationType(reference.template) },
        ]}
        evalPublished
        buttons={
          <div className="item-shortcut-group">
            <ShowIf if={!this.props.targetDoc && !this.props.readOnly}>
              <NeedAuthorization roles={['admin', 'editor']}>
                <a
                  className="item-shortcut btn btn-default btn-hover-danger delete"
                  onClick={this.deleteReference.bind(this, reference)}
                >
                  <Icon icon="trash-alt" />
                </a>
              </NeedAuthorization>
            </ShowIf>

            <ShowIf if={!this.props.targetDoc}>
              <I18NLink
                to={`/entity/${doc.get('sharedId')}`}
                onClick={e => e.stopPropagation()}
                className="item-shortcut btn btn-default"
              >
                <Icon icon="file" />
              </I18NLink>
            </ShowIf>
          </div>
        }
      />
    );
  }
}

Connection.contextTypes = {
  confirm: PropTypes.func,
};

Connection.defaultProps = {
  targetDoc: false,
};

Connection.propTypes = {
  targetDoc: PropTypes.bool,
  readOnly: PropTypes.bool,
  highlighted: PropTypes.bool.isRequired,
  active: PropTypes.bool.isRequired,
  targetRange: PropTypes.object,
  doc: PropTypes.object,
  relationTypes: PropTypes.object,
  reference: PropTypes.object.isRequired,
  highlightReference: PropTypes.func.isRequired,
  deleteReference: PropTypes.func,
  activateReference: PropTypes.func,
  selectReference: PropTypes.func,
  referencesSection: PropTypes.string,
};

const mapStateToProps = (state, ownProps) => {
  const { documentViewer } = state;
  return {
    highlighted: documentViewer.uiState.get('highlightedReference') === ownProps.reference._id,
    active: documentViewer.uiState.get('activeReference') === ownProps.reference._id,
    targetRange: documentViewer.uiState.get('reference').get('targetRange'),
    targetDoc: !!documentViewer.targetDoc.get('_id'),
    relationTypes: documentViewer.relationTypes,
    doc: selectDoc(state),
  };
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      highlightReference,
      activateReference,
      selectReference,
      deleteReference,
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(Connection);
