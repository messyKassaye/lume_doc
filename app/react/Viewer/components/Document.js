import 'app/Viewer/scss/conversion_base.scss';
import 'app/Viewer/scss/document.scss';

import PropTypes from 'prop-types';
import React, { Component } from 'react';

import Loader from 'app/components/Elements/Loader';
import PDF from 'app/PDF';
import Immutable from 'immutable';
import { highlightSnippet } from 'app/Viewer/actions/uiActions';

import determineDirection from '../utils/determineDirection';

import { APIURL } from '../../config.js';

export class Document extends Component {
  constructor(props) {
    super(props);
    this.pages = { 1: 0 };
    this.previousMostVisible = props.page;
    this.pdfLoaded = this.pdfLoaded.bind(this);
    this.onDocumentReady = this.onDocumentReady.bind(this);
    this.onTextDeselection = this.onTextDeselection.bind(this);
    this.onTextSelected = this.onTextSelected.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.highlightReference = this.highlightReference.bind(this);
  }

  componentDidMount() {
    this.props.unsetSelection();
  }

  componentDidUpdate(prevProps) {
    if (prevProps && prevProps.doc.get('_id') !== this.props.doc.get('_id')) {
      this.props.unsetSelection();
    }
    highlightSnippet(this.props.selectedSnippet, this.props.searchTerm);
  }

  onTextSelected(textSelection) {
    if (this.props.disableTextSelection) {
      return;
    }
    const selectionRectangles = textSelection.selectionRectangles.map(
      ({ regionId, ...otherProps }) => ({ ...otherProps, page: regionId })
    );
    this.props.setSelection({ ...textSelection, selectionRectangles }, this.props.file._id);
    return this.props.deactivateReference();
  }

  onTextDeselection() {
    this.props.unsetSelection();
  }

  onDocumentReady() {
    this.props.onDocumentReady(this.props.doc);
  }

  handleClick() {
    if (this.props.executeOnClickHandler) {
      this.props.onClick();
    }
  }

  highlightReference(connection) {
    return this.props.activateReference(connection, this.props.file.pdfInfo);
  }

  pdfLoaded() {
    if (this.props.doScrollToActive) {
      const references = this.props.references.toJS();
      this.props.scrollToActive(
        references.find(r => r._id === this.props.activeReference),
        this.props.file.pdfInfo,
        references,
        this.props.doScrollToActive
      );
    }

    this.componentDidUpdate();
  }

  handleOver() {}

  renderPDF(file) {
    if (!file._id) {
      return <Loader />;
    }

    return (
      <PDF
        onPageChange={this.props.onPageChange}
        onPDFReady={this.onDocumentReady}
        pdfInfo={file.pdfInfo}
        onLoad={this.pdfLoaded}
        file={`${APIURL}files/${file.filename}`}
        filename={file.filename}
        onTextSelection={this.onTextSelected}
        onTextDeselection={this.onTextDeselection}
        highlightReference={this.highlightReference}
        activeReference={this.props.activeReference}
      />
    );
  }

  render() {
    const { file } = this.props;

    const Header = this.props.header;
    return (
      <div>
        <div
          className={`_${this.props.doc.get('_id')} document ${
            this.props.className
          } ${determineDirection(file)}`}
        >
          <Header />
          <div
            className="pages"
            ref={ref => (this.pagesContainer = ref)}
            onMouseOver={this.handleOver.bind(this)}
            onClick={this.handleClick}
          >
            {this.renderPDF(file)}
          </div>
        </div>
      </div>
    );
  }
}

Document.defaultProps = {
  onDocumentReady: () => {},
  onPageChange: () => {},
  onClick: () => {},
  file: {},
  searchTerm: '',
  page: 1,
  selectedSnippet: Immutable.fromJS({}),
  deactivateReference: () => {},
  header: () => false,
  activateReference: () => {},
  doScrollToActive: false,
  scrollToActive: () => {},
  activeReference: '',
  className: '',
  executeOnClickHandler: false,
};

Document.propTypes = {
  onPageChange: PropTypes.func,
  onDocumentReady: PropTypes.func,
  doc: PropTypes.object,
  file: PropTypes.object,
  selectedSnippet: PropTypes.instanceOf(Immutable.Map),
  setSelection: PropTypes.func,
  unsetSelection: PropTypes.func,
  disableTextSelection: PropTypes.bool,
  header: PropTypes.func,
  searchTerm: PropTypes.string,
  page: PropTypes.number,
  activateReference: PropTypes.func,
  doScrollToActive: PropTypes.bool,
  scrollToActive: PropTypes.func,
  activeReference: PropTypes.string,
  references: PropTypes.object,
  className: PropTypes.string,
  onClick: PropTypes.func,
  executeOnClickHandler: PropTypes.bool,
  deactivateReference: PropTypes.func,
};

export default Document;
