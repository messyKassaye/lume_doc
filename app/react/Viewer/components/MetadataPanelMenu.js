import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { actions } from 'app/Metadata';
import { MenuButtons } from 'app/ContextMenu';
import { NeedAuthorization } from 'app/Auth';
import { Icon } from 'UI';

export class MetadataPanelMenu extends Component {
  render() {
    if (this.props.targetDoc) {
      return false;
    }
    return (
      <div>
        {(() => {
          if (this.props.docForm && this.props.docForm._id) {
            let disabled = true;
            if (this.props.formState.dirty) {
              disabled = false;
            }

            return (
              <MenuButtons.Main disabled={disabled}>
                <button type="submit" form="metadataForm" disabled={disabled}>
                  <Icon icon="save" />
                </button>
              </MenuButtons.Main>
            );
          }
          return (
            <NeedAuthorization roles={['admin', 'editor']}>
              <MenuButtons.Main
                onClick={() =>
                  this.props.loadInReduxForm(
                    'documentViewer.docForm',
                    this.props.doc.toJS(),
                    this.props.templates.toJS()
                  )
                }
              >
                <Icon icon="pencil-alt" />
              </MenuButtons.Main>
            </NeedAuthorization>
          );
        })()}
      </div>
    );
  }
}

MetadataPanelMenu.propTypes = {
  templates: PropTypes.object,
  doc: PropTypes.object,
  docForm: PropTypes.object,
  formState: PropTypes.object,
  loadInReduxForm: PropTypes.func,
  targetDoc: PropTypes.bool,
};

const mapStateToProps = ({ documentViewer, templates }) => ({
  doc: documentViewer.doc,
  templates,
  docForm: documentViewer.docForm,
  formState: documentViewer.docFormState,
  targetDoc: !!documentViewer.targetDoc.get('_id'),
});

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ loadInReduxForm: actions.loadInReduxForm }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(MetadataPanelMenu);
