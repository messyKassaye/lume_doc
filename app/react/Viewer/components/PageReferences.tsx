import React, { FunctionComponent } from 'react';
import { connect } from 'react-redux';
import { IStore } from 'app/istore';
import { ConnectionSchema } from 'shared/types/connectionType';
import { createSelector } from 'reselect';
import { Highlight } from 'react-pdf-handler';
import { unique } from 'shared/filterUnique';

export interface PageReferencesProps {
  references: { [key: string]: ConnectionSchema[] | undefined };
  page: string;
  activeReference: string;
  onClick: (c: ConnectionSchema) => {};
}

export const PageReferencesComponent: FunctionComponent<PageReferencesProps> = (
  props: PageReferencesProps
) => (
  <>
    {(props.references[props.page] || []).map((r: ConnectionSchema) => {
      const color = r._id === props.activeReference ? '#ffd84b' : '#feeeb4';

      if (!r.reference) {
        return false;
      }
      const selectionRectangles = r.reference.selectionRectangles.map(
        ({ page, ...otherProps }) => ({ regionId: page, ...otherProps })
      );
      const highlight = { ...r.reference, selectionRectangles };

      return (
        <div
          data-id={r._id}
          key={r._id?.toString()}
          className="reference"
          onClick={props.onClick.bind(null, r)}
        >
          <Highlight regionId={props.page.toString()} highlight={highlight} color={color} />
        </div>
      );
    })}
  </>
);

const indexdReferencesByPage = createSelector(
  (state: IStore) =>
    state.documentViewer.targetDocReferences.size
      ? state.documentViewer.targetDocReferences
      : state.documentViewer.references,
  (state: IStore) =>
    state.documentViewer.targetDoc.get('_id')
      ? state.documentViewer.targetDoc
      : state.documentViewer.doc,
  (references, doc) =>
    references
      .toJS()
      .reduce(
        (mappedReferences: { [key: string]: ConnectionSchema[] }, connection: ConnectionSchema) => {
          if (doc.get('sharedId') !== connection.entity) {
            return mappedReferences;
          }

          const pages = (connection.reference?.selectionRectangles || [])
            .map(selection => selection.page)
            .filter(unique);

          pages.forEach(page => {
            if (!page) {
              return;
            }

            if (!mappedReferences[page]) {
              // eslint-disable-next-line no-param-reassign
              mappedReferences[page] = [];
            }

            mappedReferences[page].push(connection);
          });

          return mappedReferences;
        },
        {}
      )
);

const mapStateToProps = (state: IStore) => ({
  references: indexdReferencesByPage(state),
  activeReference: state.documentViewer.uiState.get('activeReference'),
});

export const PageReferences = connect(mapStateToProps)(PageReferencesComponent);
