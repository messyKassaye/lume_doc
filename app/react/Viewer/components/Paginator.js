import PropTypes from 'prop-types';
import React from 'react';

import { Translate } from 'app/I18N';
import { withRouter } from 'react-router';
import { CurrentLocationLink } from 'app/Layout';

const disableButton = (page, pageToDisable) => ({
  className: page === pageToDisable ? 'btn disabled' : 'btn',
  rel: page === pageToDisable ? 'nofollow' : undefined,
});

const Paginator = ({ page, totalPages, onPageChange }) => {
  const prevPage = page - 1 || 1;
  const nextPage = page + 1 > totalPages ? totalPages : page + 1;
  return (
    <div className="paginator">
      <CurrentLocationLink
        queryParams={{ page: prevPage }}
        onClick={e => {
          e.preventDefault();
          onPageChange(prevPage);
        }}
        {...disableButton(page, 1)}
      >
        <Translate>Previous</Translate>
      </CurrentLocationLink>
      <span>{` ${page} / ${totalPages} `}</span>
      <CurrentLocationLink
        queryParams={{ page: nextPage }}
        onClick={e => {
          e.preventDefault();
          onPageChange(nextPage);
        }}
        {...disableButton(page, totalPages)}
      >
        <Translate>Next</Translate>
      </CurrentLocationLink>
    </div>
  );
};

Paginator.defaultProps = {
  page: 1,
  totalPages: 1,
  onPageChange: () => {},
};

Paginator.propTypes = {
  page: PropTypes.number,
  totalPages: PropTypes.number,
  onPageChange: PropTypes.func,
};

export default Paginator;

const PaginatorWithPage = withRouter(props => {
  const { location, ...restProps } = props;
  return <Paginator {...restProps} page={Number(location.query.page || 1)} />;
});

export { PaginatorWithPage };
