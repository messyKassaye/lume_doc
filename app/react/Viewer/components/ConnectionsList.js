import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { t } from 'app/I18N';
import { Icon } from 'UI';
import { Warning } from 'app/Layout';
import { closePanel, deactivateReference } from 'app/Viewer/actions/uiActions';
import Connection from './Connection';
import { sortTextSelections } from '../utils/sortTextSelections';

import 'app/Viewer/scss/viewReferencesPanel.scss';

export class ConnectionsList extends Component {
  static blankStateMessage(title, message) {
    return (
      <div className="blank-state">
        <Icon icon="sitemap" />
        <h4>{t('System', title)}</h4>
        <p>{t('System', message)}</p>
      </div>
    );
  }

  close() {
    this.props.closePanel();
    this.props.deactivateReference();
  }

  render() {
    const references = this.props.references
      .toJS()
      .sort((a, b) => sortTextSelections(a.reference, b.reference));

    if (this.props.loading) {
      return false;
    }

    if (!this.props.references.size && this.props.referencesSection === 'references') {
      return ConnectionsList.blankStateMessage('No References', 'No References description');
    }

    if (!this.props.references.size) {
      return ConnectionsList.blankStateMessage('No Connections', 'No Connections description');
    }

    return (
      <div className="item-group">
        {references.length > 299 && (
          <Warning inline>
            Text references are temporarily limited to a maximum of 300. We are currently working on
            a fix for this issue.
          </Warning>
        )}
        {(() =>
          references.map(reference => (
            <Connection key={reference._id} readOnly={this.props.readOnly} reference={reference} />
          )))()}
      </div>
    );
  }
}

ConnectionsList.propTypes = {
  references: PropTypes.object,
  deactivateReference: PropTypes.func,
  closePanel: PropTypes.func,
  loading: PropTypes.bool,
  readOnly: PropTypes.bool,
  referencesSection: PropTypes.string,
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ closePanel, deactivateReference }, dispatch);
}

export default connect(null, mapDispatchToProps)(ConnectionsList);
