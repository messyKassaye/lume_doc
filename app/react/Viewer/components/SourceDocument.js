import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { setSelection, unsetSelection } from 'app/Viewer/actions/selectionActions';
import {
  resetReferenceCreation,
  highlightReference,
  activateReference,
  scrollToActive,
  deactivateReference,
} from 'app/Viewer/actions/uiActions';
import Document from 'app/Viewer/components/Document';
import { createSelector } from 'reselect';
import { selectDoc, selectReferences } from '../selectors';

const selectSourceRange = createSelector(
  s => s.uiState,
  u => u.toJS().reference.sourceRange
);
const selectActiveRef = createSelector(
  s => s.uiState,
  u => u.toJS().activeReference
);

const mapStateToProps = state => {
  const { user, documentViewer } = state;
  return {
    selectedSnippet: documentViewer.uiState.get('snippet'),
    selection: selectSourceRange(documentViewer),
    doScrollToActive: documentViewer.uiState.get('goToActive'),
    doc: selectDoc(state),
    references: selectReferences(state),
    className: 'sourceDocument',
    activeReference: selectActiveRef(documentViewer),
    executeOnClickHandler: !!documentViewer.targetDoc.get('_id'),
    disableTextSelection: !user.get('_id'),
    panelIsOpen: !!documentViewer.uiState.get('panel'),
    forceSimulateSelection:
      documentViewer.uiState.get('panel') === 'targetReferencePanel' ||
      documentViewer.uiState.get('panel') === 'referencePanel',
  };
};

function mapDispatchToProps(dispatch) {
  const actions = {
    setSelection,
    unsetSelection,
    onClick: resetReferenceCreation,
    highlightReference,
    activateReference,
    scrollToActive,
    deactivateReference,
  };
  return bindActionCreators(actions, dispatch);
}

function mergeProps(stateProps, dispatchProps, ownProps) {
  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    unsetSelection: dispatchProps.unsetSelection,
  };
}
export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(Document);
