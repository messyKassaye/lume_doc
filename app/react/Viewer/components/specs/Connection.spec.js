import React from 'react';
import { shallow } from 'enzyme';
import Immutable from 'immutable';

import { Connection } from 'app/Viewer/components/Connection';
import { Item } from 'app/Layout';

describe('Connection', () => {
  let component;
  let props;

  beforeEach(() => {
    props = {
      reference: {
        _id: 'ref1',
        relationType: 'rel1',
        associatedRelationship: { entityData: { _id: '1' } },
        reference: { selectionRectangles: [], text: 'hello' },
      },
      referencedDocuments: Immutable.fromJS([
        { title: 'doc1', _id: '1' },
        { title: 'doc2', _id: '2' },
      ]),
      relationTypes: Immutable.fromJS([{ _id: 'rel1', name: 'Supports' }]),
      highlightReference: jasmine.createSpy('highlightReference'),
      activateReference: jasmine.createSpy('activateReference'),
      selectReference: jasmine.createSpy('selectReference'),
      deactivateReference: jasmine.createSpy('deactivateReference'),
      deleteReference: jasmine.createSpy('deleteReference'),
      active: true,
      highlighted: false,
      doc: { defaultDoc: { pdfInfo: '' } },
    };
  });

  const render = () => {
    component = shallow(<Connection {...props} />);
  };

  it('should render the reference', () => {
    render();
    expect(component).toMatchSnapshot();
  });

  it('should disable all-document connections if targetDocument', () => {
    props.targetDoc = true;
    delete props.reference.reference;
    render();

    expect(component.find(Item).get(0).props.className).toContain('disabled');
  });

  describe('when mouseenter on a reference', () => {
    it('should highlightReference', () => {
      render();
      component
        .find(Item)
        .last()
        .simulate('mouseenter');
      expect(props.highlightReference).toHaveBeenCalledWith('ref1');
    });
  });

  describe('when mouseleave a reference', () => {
    it('should unhighlightReference', () => {
      render();
      component
        .find(Item)
        .last()
        .simulate('mouseleave');
      expect(props.highlightReference).toHaveBeenCalledWith(null);
    });
  });

  describe('when click on a reference', () => {
    beforeEach(() => {
      props.uiState = Immutable.fromJS({
        reference: Immutable.fromJS({ targetRange: 'targetRange' }),
        panel: 'ConnectionsList',
        activeReference: 'ref1',
      });
      props.referencesSection = 'tabName';
    });

    describe('when document is source document', () => {
      it('should activate it', () => {
        render();
        component.find(Item).simulate('click');
        expect(props.activateReference).toHaveBeenCalledWith(
          props.reference,
          props.doc.defaultDoc.pdfInfo,
          'tabName'
        );
        expect(component.find(Item).props().className).toContain('relationship-active');
      });

      describe('when readOnly', () => {
        it('should not activate it', () => {
          props.readOnly = true;
          render();
          component
            .find(Item)
            .last()
            .simulate('click');
          expect(props.activateReference).not.toHaveBeenCalled();
          expect(component.find(Item).props().className).not.toContain('relationship-active');
        });
      });
    });

    describe('when document is target document', () => {
      it('should select it', () => {
        props.targetDoc = true;
        props.targetRange = {};
        render();
        component
          .find(Item)
          .last()
          .simulate('click');
        expect(props.selectReference).toHaveBeenCalledWith(props.reference, '');
        expect(
          component
            .find(Item)
            .last()
            .getElements()[0].props.className
        ).toContain('relationship-selected');
      });

      describe('when connection is to the entire document', () => {
        it('should not select it', () => {
          props.targetDoc = true;
          delete props.reference.reference;
          render();
          component
            .find(Item)
            .first()
            .simulate('click');
          expect(props.selectReference).not.toHaveBeenCalled();
        });
      });
    });
  });
});
