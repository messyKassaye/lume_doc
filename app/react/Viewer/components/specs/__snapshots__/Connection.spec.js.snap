// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connection should render the reference 1`] = `
<Connect(Item)
  additionalMetadata={
    Array [
      Object {
        "label": "Connection type",
        "value": undefined,
      },
    ]
  }
  additionalText={null}
  buttons={
    <div
      className="item-shortcut-group"
    >
      <ShowIf
        if={true}
      >
        <Connect(NeedAuthorization)
          roles={
            Array [
              "admin",
              "editor",
            ]
          }
        >
          <a
            className="item-shortcut btn btn-default btn-hover-danger delete"
            onClick={[Function]}
          >
            <Connect(Icon)
              icon="trash-alt"
            />
          </a>
        </Connect(NeedAuthorization)>
      </ShowIf>
      <ShowIf
        if={true}
      >
        <Connect(I18NLink)
          className="item-shortcut btn btn-default"
          onClick={[Function]}
          to="/entity/undefined"
        >
          <Connect(Icon)
            icon="file"
          />
        </Connect(I18NLink)>
      </ShowIf>
    </div>
  }
  className="relationship-active item-ref1  "
  data-id="ref1"
  doc={
    Immutable.Map {
      "_id": "1",
    }
  }
  evalPublished={true}
  noMetadata={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
/>
`;
